### PR TITLE
MULE-15430: Add Value Provider support for email folders

### PR DESCRIPTION
**Pull request by Bardammu**

### CODE REVIEW

File: src/main/java/org/mule/extension/email/internal/value/MailboxFolderValueProvider.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @since 1.1.0
+ */
+public class MailboxFolderValueProvider implements ValueProvider {
+
+  @Connection
+  private MailboxConnection mailboxConnection;
+
+  @Override
+  public Set<Value> resolve() throws ValueResolvingException {
+    Folder[] folders = mailboxConnection.listFolders();
```
Comment (estebanwasinger) [inline src/main/java/org/mule/extension/email/internal/value/MailboxFolderValueProvider.java]: You may catch `EmailAccessingFolderException` and throw a `ValueResolvingException`

File: src/main/java/org/mule/extension/email/internal/value/MailboxFolderValueProvider.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    List<String> foldersNames = new ArrayList<>();
```
Comment (estebanwasinger) [inline src/main/java/org/mule/extension/email/internal/value/MailboxFolderValueProvider.java]: You could simplify your code with something more functional: `return getValuesFor(stream(folders).map(Folder::getFullName));`

File: src/main/java/org/mule/extension/email/internal/mailbox/MailboxConnection.java [hunk 3]
```diff
@@ -134,6 +134,21 @@ public synchronized Folder getFolder(String mailBoxFolder, int openMode) {
...
 
+  /**
+   * Retrieves all the folders of the user's personal namespace.
+   *
+   * @return the list of {@link Folder}.
+   */
+  public synchronized Folder[] listFolders() {
+    Folder[] folders;
+    try {
+      folders = store.getDefaultFolder().list("*");
```
Comment (estebanwasinger) [inline src/main/java/org/mule/extension/email/internal/mailbox/MailboxConnection.java]: just `return store.getDefaultFolder().list("*");`

File: src/test/java/org/mule/extension/email/value/MailboxFolderValueProviderTestCase.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  private static final String SPECIAL_CHARACTER_PASSWORD = "*uawH*IDXlh2p%21xSPOx%23%25zLpL";
+  private static final String SALESFORCE_FOLDER = "salesforce";
+  private static final String MULESOFT_FOLDER = SALESFORCE_FOLDER + "/mulesoft";
+
+  @Rule
+  public SystemProperty specialCharacterPassword = new SystemProperty("specialCharacterPassword", SPECIAL_CHARACTER_PASSWORD);
+
+  @Inject
+  private ValueProviderService service;
```
Comment (estebanwasinger) [inline src/test/java/org/mule/extension/email/value/MailboxFolderValueProviderTestCase.java]: To have a more authentic execution, a more similar to what happens in Studio, Value Provider/ Metadata tests should run on Lazy Mode that is done overriding some methods:

```java
  @Override
  public boolean enableLazyInit() {
    return true;
  }

  @Override
  public boolean disableXmlValidations() {
    return true;
  }
```

Also you will need to do some tunning on how to importe the `ValueProviderService`:

```java
@Inject
@Named(VALUE_PROVIDER_SERVICE_KEY)
private ValueProviderService service;
```

**Comment by Bardammu** [comment]
screen shot 2018-09-03 at 15 28 15

**Pull request merged**